src/ripple/app/main/GRPCServer.cpp

```diff
@@ -0,0 +1,408 @@
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/app/main/GRPCServer.h>
+#include <ripple/resource/Fees.h>
+
+namespace ripple {
+
+namespace {
+
+// helper function. strips scheme from endpoint string
+std::string
+getEndpoint(std::string const& peer)
+{
+    std::size_t first = peer.find_first_of(":");
+    std::size_t last = peer.find_last_of(":");
+    std::string peerClean(peer);
+    if (first != last)
+    {
+        peerClean = peer.substr(first + 1);
+    }
+    return peerClean;
+}
+}  // namespace
+
+template <class Request, class Response>
+GRPCServerImpl::CallData<Request, Response>::CallData(
+    rpc::v1::XRPLedgerAPIService::AsyncService& service,
+    grpc::ServerCompletionQueue& cq,
+    Application& app,
+    BindListener<Request, Response> bindListener,
+    Handler<Request, Response> handler,
+    RPC::Condition requiredCondition,
+    Resource::Charge loadType)
+    : service_(service)
+    , cq_(cq)
+    , finished_(false)
+    , app_(app)
+    , aborted_(false)
+    , responder_(&ctx_)
+    , bindListener_(std::move(bindListener))
+    , handler_(std::move(handler))
+    , requiredCondition_(std::move(requiredCondition))
+    , loadType_(std::move(loadType))
+{
+    // Bind a listener. When a request is received, "this" will be returned
+    // from CompletionQueue::Next
+    bindListener_(service_, &ctx_, &request_, &responder_, &cq_, &cq_, this);
+}
+
+template <class Request, class Response>
+std::shared_ptr<Processor>
+GRPCServerImpl::CallData<Request, Response>::clone()
+{
+    return std::make_shared<CallData<Request, Response>>(
+        service_,
+        cq_,
+        app_,
+        bindListener_,
+        handler_,
+        requiredCondition_,
+        loadType_);
+}
+
+template <class Request, class Response>
+void
+GRPCServerImpl::CallData<Request, Response>::process()
+{
+    // sanity check
+    BOOST_ASSERT(!finished_);
+
+    std::shared_ptr<CallData<Request, Response>> thisShared =
+        this->shared_from_this();
+    app_.getJobQueue().postCoro(
+        JobType::jtRPC,
+        "gRPC-Client",
+        [thisShared](std::shared_ptr<JobQueue::Coro> coro) {
+            std::lock_guard<std::mutex> lock(thisShared->mut_);
```

Comment:
Stylistic suggestion:  Change this to:
```
std::lock_guard lock{thisShared->mut_};
```
Rationale: CTAD makes it easier for us to change the type of the `mutex` if we want in the future, because we don't have to change it here too.

The rationale for using `{}` over `()` is weaker.  I like it because I don't have to wonder if this is a function call or a constructor.  But I'm not loosing sleep over it either way.

Ditto for the two other places in this file that use `lock_guard`.